### ISSUE TITLE
Allow sub-test selection

### ISSUE THREAD

**Issue by smyrman** [microsoft/vscode-go#2850]
If you have a question, please ask it on https://gitter.im/Microsoft/vscode-go or on the VSCode channel in Gophers Slack(https://invite.slack.golangbridge.org/)

**Is your feature request related to a problem? Please describe.**

VS-Code has a _code lens_ link that allows you to run individual tests. 

![image](https://user-images.githubusercontent.com/75205/67205997-20988a80-f411-11e9-9786-afa3ac1883e7.png)

This works by generating a paramater ala `-run ^(TestDeepEquals)$` for the `go test` command. This is great, but for large tests and table tests, it would be great if there was also a possibility to focus on a single _sub-test_ (as [introduced](https://blog.golang.org/subtests) in Go 1.7). This is something that Go already supports by providing a flag ala  `-run ^(TestDeepEquals/sub-test)$`.

A method of doing this has been suggested before in #1748, but closed due to the complexity of the solution. I want to introduce a hopefully much simpler solution to the problem.

**Describe the solution you'd like**

 I see two possible solutions, which would ideally be _combined_.

Solution 1:

In the test failure output, allow clicking on a failing sub-test to rerun it with the right "name" added to the `-run` parameter.

Solution 2 alternative A:

In the code lense, add a link to run a sub-test. I am a bit on thin ice here, but since go test results are _cached_, if we could easily access this cache, perhaps it's possible to issue a dialog with auto-complete to search for (and select) any previously run sub-test. If no sub-tests have previously been run, the dialog could either be hidden, or allow free-text typing.

![image](https://user-images.githubusercontent.com/75205/67206645-5e49e300-f412-11e9-8700-26a5ade5db7a.png)

Solution 2 alternative B:

As a variant of solution 2, the drop-down could provide only _failing sub-tests_ from the previous run of the full test (that VS Code knows about).

**Describe alternatives you've considered**

Solution #1748 have previously been suggested with a different solution.

**UPDATE**: Replaces wrong link to previous issue (the issue is 1748, not 597)

**Comment by ramya-rao-a** [comment]
Thanks for reporting @smyrman 

#597 was related to generating of tests, where as I believe what you are asking for is running of the sub tests, did I get that right?

**Comment by ramya-rao-a** [comment]
Can you provide a small sample file with subtests that I can use to check the current behavior?

**Comment by smyrman** [comment]
Sorry, when navigating back and forth I managed to isnert the wrong sub-test related issue link. 
The intended linked issue was #1748, not 597. Will update the description with this.

**Comment by smyrman** [comment]
> Can you provide a small sample file with subtests that I can use to check the current behavior?

A straight forward failing sub-test would be.

```go
pacage pkg_test

import "testing"

func TestFoo(t *testing.T) {
    t.Run("foo", func(t *testing.T) {
        t.Fatalf("foo failed")
    }
    t.Run("bar", func(t *testing.T) {
        t.Fatalf("bar failed")
    }
}
````
Then, be able to run either just `TestFoo/foo` or `TestFoo/bar`, preferably without typing the whole name.

Typing the whole name gets inconvinient as the sub-test name gets longer, such as for instance here:
- https://github.com/searis/subtest/blob/3242f2e96b8a554934c49f1f14f8a127e9e5ab07/check_test.go#L11

**Comment by ramya-rao-a** [comment]
Thanks for that clarification

> Solution 1: In the test failure output, allow clicking on a failing sub-test to rerun it with the right "name" added to the -run parameter.

The output pane only allows links that take you to a particular file or an external link. As far as I know, having a link that do custom action on clicking is not a feature supported in VS Code

> Solution 2  alternative A:
In the code lense, add a link to run a sub-test. I am a bit on thin ice here, but since go test results are cached, if we could easily access this cache, perhaps it's possible to issue a dialog with auto-complete to search for (and select) any previously run sub-test.

The extension doesnt access the cache. Generally, we have refrained from any such tasks that assume the knowledge of how the Go command works. We depend on Go tools in the tooling ecosystem to interact and do most of the heavy lifting. So, as I mentioned in #1748, if there is a Go tool that allows us to identify/list the sub tests, then we can use that.

> Solution 2 alternative B:
As a variant of solution 2, the drop-down could provide only failing sub-tests from the previous run of the full test (that VS Code knows about).

This would require the extension to parse the test output to determine the failing tests. If we go down this route, then I would expect the feature to be able to run any of the failed tests from the previous run and not just the sub tests.

We have a command `Go: Test Previous` that re-runs the last test command.
We could have a command `Go: Test Previous failed` that does the above.

But, this is not trivial as we have to parse the test output which in my experience has never been straight forward :)

If someone wants to give it a try, they are most welcome.

**Comment by smyrman** [comment]
For solution 2 alternative B, the output from `go test -json` is probably not too hard to parse. 

Here is an example:

Given the following test file `play/footest/pkg_test.go`:
```go
package pkg_test

import "testing"

package pkg_test

import "testing"

func TestFoo(t *testing.T) {
	t.Run("foo", func(t *testing.T) {
		t.Fatalf("foo failed")
	})
	t.Run("bar", func(t *testing.T) {
		t.Fatalf("bar failed")
	})
}

func TestBar(t *testing.T) {
	t.Run("foo", func(t *testing.T) {
		t.Logf("foo passed")
	})
	t.Run("bar", func(t *testing.T) {
		t.Errorf("bar error")
		t.Fatalf("bar failed")
	})
}

func TestBar(t *testing.T) {
	t.Run("foo", func(t *testing.T) {
		t.Logf("foo passed")
	})
	t.Run("bar", func(t *testing.T) {
		t.Fatalf("bar failed")
	})
}
```
Output of `go test -json`:
```
{"Time":"2019-11-05T11:21:26.17628+01:00","Action":"run","Package":"play/footest","Test":"TestFoo"}
{"Time":"2019-11-05T11:21:26.176588+01:00","Action":"output","Package":"play/footest","Test":"TestFoo","Output":"=== RUN   TestFoo\n"}
{"Time":"2019-11-05T11:21:26.176601+01:00","Action":"run","Package":"play/footest","Test":"TestFoo/foo"}
{"Time":"2019-11-05T11:21:26.176609+01:00","Action":"output","Package":"play/footest","Test":"TestFoo/foo","Output":"=== RUN   TestFoo/foo\n"}
{"Time":"2019-11-05T11:21:26.176623+01:00","Action":"run","Package":"play/footest","Test":"TestFoo/bar"}
{"Time":"2019-11-05T11:21:26.17663+01:00","Action":"output","Package":"play/footest","Test":"TestFoo/bar","Output":"=== RUN   TestFoo/bar\n"}
{"Time":"2019-11-05T11:21:26.176647+01:00","Action":"output","Package":"play/footest","Test":"TestFoo","Output":"--- FAIL: TestFoo (0.00s)\n"}
{"Time":"2019-11-05T11:21:26.176786+01:00","Action":"output","Package":"play/footest","Test":"TestFoo/foo","Output":"    --- FAIL: TestFoo/foo (0.00s)\n"}
{"Time":"2019-11-05T11:21:26.176839+01:00","Action":"output","Package":"play/footest","Test":"TestFoo/foo","Output":"        pkg_test.go:7: foo failed\n"}
{"Time":"2019-11-05T11:21:26.176856+01:00","Action":"fail","Package":"play/footest","Test":"TestFoo/foo","Elapsed":0}
{"Time":"2019-11-05T11:21:26.176874+01:00","Action":"output","Package":"play/footest","Test":"TestFoo/bar","Output":"    --- FAIL: TestFoo/bar (0.00s)\n"}
{"Time":"2019-11-05T11:21:26.176879+01:00","Action":"output","Package":"play/footest","Test":"TestFoo/bar","Output":"        pkg_test.go:10: bar failed\n"}
{"Time":"2019-11-05T11:21:26.176883+01:00","Action":"fail","Package":"play/footest","Test":"TestFoo/bar","Elapsed":0}
{"Time":"2019-11-05T11:21:26.176887+01:00","Action":"fail","Package":"play/footest","Test":"TestFoo","Elapsed":0}
{"Time":"2019-11-05T11:21:26.176891+01:00","Action":"run","Package":"play/footest","Test":"TestBar"}
{"Time":"2019-11-05T11:21:26.176921+01:00","Action":"output","Package":"play/footest","Test":"TestBar","Output":"=== RUN   TestBar\n"}
{"Time":"2019-11-05T11:21:26.176931+01:00","Action":"run","Package":"play/footest","Test":"TestBar/foo"}
{"Time":"2019-11-05T11:21:26.176938+01:00","Action":"output","Package":"play/footest","Test":"TestBar/foo","Output":"=== RUN   TestBar/foo\n"}
{"Time":"2019-11-05T11:21:26.176945+01:00","Action":"run","Package":"play/footest","Test":"TestBar/bar"}
{"Time":"2019-11-05T11:21:26.176951+01:00","Action":"output","Package":"play/footest","Test":"TestBar/bar","Output":"=== RUN   TestBar/bar\n"}
{"Time":"2019-11-05T11:21:26.176958+01:00","Action":"output","Package":"play/footest","Test":"TestBar","Output":"--- FAIL: TestBar (0.00s)\n"}
{"Time":"2019-11-05T11:21:26.176966+01:00","Action":"output","Package":"play/footest","Test":"TestBar/foo","Output":"    --- PASS: TestBar/foo (0.00s)\n"}
{"Time":"2019-11-05T11:21:26.176973+01:00","Action":"output","Package":"play/footest","Test":"TestBar/foo","Output":"        pkg_test.go:16: foo passed\n"}
{"Time":"2019-11-05T11:21:26.177011+01:00","Action":"pass","Package":"play/footest","Test":"TestBar/foo","Elapsed":0}
{"Time":"2019-11-05T11:21:26.177021+01:00","Action":"output","Package":"play/footest","Test":"TestBar/bar","Output":"    --- FAIL: TestBar/bar (0.00s)\n"}
{"Time":"2019-11-05T11:21:26.177027+01:00","Action":"output","Package":"play/footest","Test":"TestBar/bar","Output":"        pkg_test.go:19: bar error\n"}
{"Time":"2019-11-05T11:21:26.177034+01:00","Action":"output","Package":"play/footest","Test":"TestBar/bar","Output":"        pkg_test.go:20: bar failed\n"}
{"Time":"2019-11-05T11:21:26.177044+01:00","Action":"fail","Package":"play/footest","Test":"TestBar/bar","Elapsed":0}
{"Time":"2019-11-05T11:21:26.17705+01:00","Action":"fail","Package":"play/footest","Test":"TestBar","Elapsed":0}
{"Time":"2019-11-05T11:21:26.177067+01:00","Action":"output","Package":"play/footest","Output":"FAIL\n"}
{"Time":"2019-11-05T11:21:26.177334+01:00","Action":"output","Package":"play/footest","Output":"FAIL\tplay/footest\t0.006s\n"}
{"Time":"2019-11-05T11:21:26.177362+01:00","Action":"fail","Package":"play/footest","Elapsed":0.007}
```
Notice that each line is a separate JSON document, and there is no surrounding square brackets `[]`.

Output of `go test .`:
```
--- FAIL: TestFoo (0.00s)
    --- FAIL: TestFoo/foo (0.00s)
        pkg_test.go:7: foo failed
    --- FAIL: TestFoo/bar (0.00s)
        pkg_test.go:10: bar failed
--- FAIL: TestBar (0.00s)
    --- FAIL: TestBar/bar (0.00s)
        pkg_test.go:19: bar error
        pkg_test.go:20: bar failed
FAIL
exit status 1
FAIL	play/footest	0.006s
```

We can find the following fields in the output:
- Package: The package being tested
- Test: Full name of the test
- Action: "output" / "run" / "fail" / "pass"
- Output: only when action is "output"
- Elapsed: only when action is "fail" or "pass".

To get a list of all tests we can:
- Look at all actions with Action "run", extract "Package" and "Test".

To get a list of all failed tests we can:
- Look at all action with Action "fail", extract "Package" and "Test".

To forward the "standard" output to stdout we can:
- For all failed tests, print the content of `Output` for all `"output"` events.

The latter can also be done via a tool, such as gotestsum --format 'standard-quiet':
- https://github.com/gotestyourself/gotestsum

**Comment by ramya-rao-a** [comment]
Apologies for the late response, but PRs are most welcome to parse the test output, identify failed tests and provide a command to re-run just the failed tests.

Code Pointers:
- [Set up local dev environment for the extension](https://github.com/microsoft/vscode-go/wiki/Building,-Debugging-and-Sideloading-the-extension-in-Visual-Studio-Code#building-and-debugging-the-extension)
- [goTest()](https://github.com/microsoft/vscode-go/blob/0.12.0/src/testUtils.ts#L195) is the function that calls `go test` and parses the output. It is here that you will need to keep track of the failed test names and return them in the rejected promise.
- Use the code for the `Go: Test Previous` command at https://github.com/microsoft/vscode-go/blob/0.12.0/src/goTest.ts#L226-L237 as a reference to have a new command `Go: Test Previously failed tests`. 
- You will need a copy of the [latestTestConfig](https://github.com/microsoft/vscode-go/blob/master/src/goTest.ts#L230) which has its [functions](https://github.com/microsoft/vscode-go/blob/0.12.0/src/testUtils.ts#L51) updated to be reflect the name of the failed tests.

**Comment by smyrman** [comment]
Is there anyway that we can perform som sort of drop-down selection form the sub-test list? I think if that is to be useful, that must be possible, or the output would look the same as if we rerun all sub-tests (also passed ones) in non-verbose mode.

**Comment by ramya-rao-a** [comment]
If you update the [functions](https://github.com/microsoft/vscode-go/blob/0.12.0/src/testUtils.ts#L51) property of the last used test config with the failed tests (in your case say `TestFoo/foo`), then the next run would be only for failed test runs. So, the passed ones will not be run.

If you want to show the failed test runs in a dropdown, you can use the `showQuickPick` method (look for it in the code, its been used in a few places) to show the drop down.

**Comment by ramya-rao-a** [comment]
Hello all,

We are in the midst of a repo move, see [We are moving](https://github.com/microsoft/vscode-go#we-are-moving---) section in our readme for more details.

Closing this issue in favor of https://github.com/golang/vscode-go/issues/122. Please subscribe to it for further updates and upvote it as well

Thanks for all the support & Happy Coding!
